source/common/network/io_uring_socket_handle_impl.cc

```diff
@@ -0,0 +1,471 @@
+#include "source/common/network/io_uring_socket_handle_impl.h"
+
+#include "envoy/buffer/buffer.h"
+#include "envoy/common/exception.h"
+#include "envoy/event/dispatcher.h"
+
+#include "source/common/api/os_sys_calls_impl.h"
+#include "source/common/buffer/buffer_impl.h"
+#include "source/common/common/assert.h"
+#include "source/common/common/utility.h"
+#include "source/common/io/io_uring.h"
+#include "source/common/network/address_impl.h"
+#include "source/common/network/io_socket_error_impl.h"
+
+namespace Envoy {
+namespace Network {
+
+namespace {
+
+constexpr socklen_t udsAddressLength() { return sizeof(sa_family_t); }
+
+} // namespace
+
+IoUringSocketHandleImpl::IoUringSocketHandleImpl(const uint32_t read_buffer_size,
+                                                 const Io::IoUringFactory& io_uring_factory,
+                                                 os_fd_t fd, bool socket_v6only,
+                                                 absl::optional<int> domain)
+    : read_buffer_size_(read_buffer_size), io_uring_factory_(io_uring_factory), fd_(fd),
+      socket_v6only_(socket_v6only), domain_(domain) {}
+
+IoUringSocketHandleImpl::~IoUringSocketHandleImpl() {
+  if (SOCKET_VALID(fd_)) {
+    // The TLS slot has been shut down by this moment with IoUring wiped out, thus
+    // better use this posix system call instead of IoUringSocketHandleImpl::close().
+    ::close(fd_);
+  }
+}
+
+Api::IoCallUint64Result IoUringSocketHandleImpl::close() {
+  ASSERT(SOCKET_VALID(fd_));
+  auto req = new Request{absl::nullopt, RequestType::Close};
+  Io::IoUringResult res = io_uring_factory_.getOrCreate().prepareClose(fd_, req);
+  if (res == Io::IoUringResult::Failed) {
+    // Fall back to posix system call.
+    ::close(fd_);
+  }
+  if (isLeader()) {
+    io_uring_factory_.getOrCreate().unregisterEventfd();
+    file_event_adapter_.reset();
```

Comment:
If the adapter deconstructs since the leading handle closes, what will happen to the remaining existing handles?